long-running-tasks/src/test/java/io/streamnative/pulsar/recipes/task/Soak.java

```diff
@@ -0,0 +1,398 @@
+
+  private PulsarClient client;
+  private Producer<Task> producer;
+  private Consumer<TaskMetadata> consumer;
+
+  private void createResources(String taskTopic) throws Exception {
+    client = PulsarClient.builder().serviceUrl(pulsar.getPulsarBrokerUrl()).build();
+    producer = client.newProducer(taskSchema).topic(taskTopic).enableBatching(false).create();
+    consumer =
+        client
+            .newConsumer(metadataSchema)
+            .topic(taskTopic + "-state")
+            .subscriptionName(randomUUID().toString())
+            .subscribe();
+  }
+
+  @AfterEach
+  void afterEach() throws Exception {
+    producer.close();
+    consumer.close();
+  }
+
+  @Test
+  @Timeout(value = 4, unit = HOURS)
+  void soak() throws Exception {
+    // Modify these parameters to match your expected workload.
+    var soakTimeout = Duration.ofHours(4);
+    var workerCount = 15;
+    var keepAliveInterval = Duration.ofSeconds(1);
+    var taskCount = 100;
+    var meanTaskTime = Duration.ofMinutes(5);
+    var maxRetries = 3;
+    var taskDurationDistribution =
+        new NormalDistribution(meanTaskTime.toSeconds(), meanTaskTime.toSeconds() / 2.0);
+    var taskOutcomeDistribution =
+        new EnumeratedDistribution<>(
+            List.of(
+                Pair.create(RESULT, 0.9),
+                Pair.create(EXCEPTION, 0.75),
+                Pair.create(TIMEOUT, 0.25)));
+
+    // Begin test
+    var taskTopic = randomUUID().toString();
+    createResources(taskTopic);
+
+    var configuration =
+        TaskWorkerConfiguration.builder(taskSchema, resultSchema)
+            .taskTopic(taskTopic)
+            .subscription("subscription")
+            .retention(Duration.ofSeconds(1))
+            .keepAliveInterval(keepAliveInterval)
+            .maxTaskAttempts(maxRetries)
+            .build();
+
+    // Start workers
+    //
+    var process = new DummyProcess(clock);
+    var workers =
+        IntStream.range(0, workerCount)
+            .mapToObj(i -> newTaskWorker(process, pulsar.getPulsarBrokerUrl(), configuration))
+            .collect(toUnmodifiableList());
+
+    // Submit work
+    //
+    var tasks =
+        new TaskFactory(maxRetries, taskDurationDistribution, taskOutcomeDistribution)
+            .newTasks(taskCount);
+
+    var unscheduledTasks =
+        tasks.stream().map(this::send).filter(Optional::isPresent).collect(toUnmodifiableSet());
+    var scheduledTasks =
+        tasks.stream().filter(t -> !unscheduledTasks.contains(t)).collect(toUnmodifiableList());
+    var cumulativeTaskTries = tasks.stream().mapToLong(t -> t.getTryOutcomes().size()).sum();
+    scheduledTasks.forEach(t -> log.debug("Task: {}", t));
+
+    log.info("Successfully scheduled {} tasks out of {}", scheduledTasks.size(), tasks.size());
+    log.info(
+        "Total expected task tries: {} (exception: {}, timeout: {}. success: {})",
+        cumulativeTaskTries,
+        tasks.stream()
```

Comment:
There's the same loop three times. Is there a better way that loops once while maintaining the conciseness of the code?

Reply:
Perhaps something like:

```
Map<Outcome, Long> expected = tasks.stream().flatMap(t -> t.getTryOutcomes().stream())
            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
```